src/mesh/topology.rs

```diff
@@ -19,43 +12,74 @@ use super::{Face, Mesh};
 /// contain before it spills into heap. Implementation detail.
 const MAX_INLINE_NEIGHBOR_COUNT: usize = 8;
 
+pub type VertexToFaceRelation = SmallVec<[u32; MAX_INLINE_NEIGHBOR_COUNT]>;
+
+/// Calculates topological relations (neighborhood) of mesh vertex -> faces.
+/// Returns a Map (key: vertex index, value: list of its neighboring faces indices)
+pub fn calculate_vertex_to_face_topology(mesh: &Mesh) -> Vec<VertexToFaceRelation> {
+    calculate_vertex_to_face_topology_from_components(mesh.faces(), cast_u32(mesh.vertices().len()))
+}
+
+pub fn calculate_vertex_to_face_topology_from_components(
+    faces: &[Face],
+    vertex_count: u32,
+) -> Vec<VertexToFaceRelation> {
+    let mut v2f: Vec<SmallVec<[u32; 8]>> = vec![SmallVec::new(); cast_usize(vertex_count)];
+
+    for (face_index, face) in faces.iter().enumerate() {
+        let face_index_u32 = cast_u32(face_index);
+
+        match face {
+            Face::Triangle(triangle_face) => {
+                let vertices = &triangle_face.vertices;
+
+                for from_vertex in &[vertices.0, vertices.1, vertices.2] {
+                    if !v2f[cast_usize(*from_vertex)].contains(&face_index_u32) {
+                        v2f[cast_usize(*from_vertex)].push(face_index_u32);
+                    }
+                }
+            }
+        }
+    }
+
+    v2f
+}
+
 /// A topology containing neighborhood relations between faces. Two
 /// faces are neighbors if and only if they they share an unoriented
 /// edge.
-pub type FaceToFaceTopology = HashMap<u32, SmallVec<[u32; MAX_INLINE_NEIGHBOR_COUNT]>>;
+pub type FaceToFaceRelation = SmallVec<[u32; MAX_INLINE_NEIGHBOR_COUNT]>;
```

Comment:
This change invalidated the comment above it. Let's keep the code and comments in sync - either revert the change or update the comment. I have a slight preference to keeping the type definition as it was (with the exception of changing the hashmap to a vector). Also for `VertexToVertexTopology` below, and ideally also for `VertexToFaceRelation` above.

The reason for leaving the types as they were is because they were placeholders for the actual wrapper structs mentioned in the above fixme you removed. All we will have to do when we implement the fixme is substitute the type alias for the wrapper. Ideally very little caller call would have to be modified (the exception being if the caller did not yet use the type alias), but the topology would become more typesafe.

Reply:
Yes I agree and I originally had it like that but Clippy made me do it this way. The reason is you can't have a Vector as a function parameter.

Reply:
Hmm. In this case I would define the type aliases as they were and not use them as function parameters yet (so clippy is satisfied). Alternatively I would add clippy allow attributes with fixmes to remove them, but that seems more complex.

Reply:
> In this case I would define the type aliases as they were and not use them as function parameters yet

Why is that a good solution? It doesn't give the caller (or linter) any hint, so why not keeping them unwrapped?
Also why is it better than the status quo?

Reply:
I am trying to figure out a way of this working without creating the full-blown wrappers right now. Yes, as you are saying, as a 3rd alternative, we can also merge the comments on the type aliases with the functions that create them and delete the type aliases for now. They will come back as the wrappers at some point.

Reply:
going for the 3rd option. the comments are already there